### PR TITLE
support contains operator

### PR DESCRIPTION
**Pull request by swilly22**
#579

### CODE REVIEW

File: src/arithmetic/arithmetic_expression.c [hunk 1]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
 	return trimmed;
 }
 
+SIValue AR_CONTAINS(SIValue *argv, int argc) {
+	assert(argc == 2);
+
+	// No string contains null.
+	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_BoolVal(false);
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: False or null?

File: src/arithmetic/arithmetic_expression.c [hunk 2]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
...
+
+	// TODO: remove once we have runtime error handling.
+	assert((SI_TYPE(argv[0]) & SI_STRING) && (SI_TYPE(argv[1]) & SI_STRING));
+
+	const char *hay = argv[0].stringval;
+	const char *needle = argv[1].stringval;
+
+	// See if needle is in hay.
+	bool found = (strstr(hay, needle) != NULL);
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: As much as it made me smile... I think we should use more conventional names as ”lookupValue” and ”lookupString” or ”lookupArray”

File: src/arithmetic/arithmetic_expression.c [hunk 3]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
...
+	return SI_BoolVal(found);
+}
+
+SIValue AR_STARTSWITH(SIValue *argv, int argc) {
+	assert(argc == 2);
+
+	// No string contains null.
+	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_BoolVal(false);
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: Verify null vs false

File: src/arithmetic/arithmetic_expression.c [hunk 4]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
...
+	}
+
+	return SI_BoolVal(true);
+}
+
+SIValue AR_ENDSWITH(SIValue *argv, int argc) {
+	assert(argc == 2);
+
+	// No string contains null.
+	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_BoolVal(false);
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: Verify null vs false

File: src/arithmetic/arithmetic_expression.c [hunk 5]
```diff
@@ -1010,6 +1085,10 @@ void AR_RegisterFuncs() {
 	AR_RegFunc(lower_func_name, lower_func_name_len, AR_TRIM);
 	lower_func_name_len = 32;
 
+	_toLower("contains", &lower_func_name[0], &lower_func_name_len);
+	AR_RegFunc(lower_func_name, lower_func_name_len, AR_CONTAINS);
+	lower_func_name_len = 32;
+
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: No AR_STARTSWITH, AR_ENDSWITH registration?

Comment (swilly22) [reply src/arithmetic/arithmetic_expression.c]: Good catch!

File: src/arithmetic/arithmetic_expression.c (hunk 4 above)
Comment (swilly22) [reply src/arithmetic/arithmetic_expression.c]: ?

File: src/arithmetic/arithmetic_expression.c (hunk 1 above)
Comment (swilly22) [reply src/arithmetic/arithmetic_expression.c]: TCK specifies false.

File: src/arithmetic/arithmetic_expression.c [hunk 6]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
...
+	return SI_BoolVal(true);
+}
+
+SIValue AR_ENDSWITH(SIValue *argv, int argc) {
+	assert(argc == 2);
+
+	// No string contains null.
+	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_BoolVal(false);
+
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: Assert string?

File: src/arithmetic/arithmetic_expression.c [hunk 7]
```diff
@@ -847,6 +859,69 @@ SIValue AR_TRIM(SIValue *argv, int argc) {
...
+	bool found = (strstr(hay, needle) != NULL);
+	return SI_BoolVal(found);
+}
+
+SIValue AR_STARTSWITH(SIValue *argv, int argc) {
+	assert(argc == 2);
+
+	// No string contains null.
+	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_BoolVal(false);
+
```
Comment (DvirDukhan) [inline src/arithmetic/arithmetic_expression.c]: Assert string?

File: src/filter_tree/filter_tree.c [hunk 8]
```diff
@@ -85,36 +85,70 @@ Vector *FilterTree_SubTrees(const FT_FilterNode *root) {
...
+		argv[0] = *aVal;
+		argv[1] = *bVal;
+		contains = AR_STARTSWITH(argv, 2);
+		return contains.longval == true;
+	case OP_ENDSWITH:
+		// String matching.
+		argv[0] = *aVal;
+		argv[1] = *bVal;
+		contains = AR_ENDSWITH(argv, 2);
+		return contains.longval == true;
```
Comment (jeffreylovitz) [inline src/filter_tree/filter_tree.c]: `contains.boolval`

File: src/filter_tree/filter_tree.c [hunk 9]
```diff
@@ -85,36 +85,70 @@ Vector *FilterTree_SubTrees(const FT_FilterNode *root) {
...
+		argv[0] = *aVal;
+		argv[1] = *bVal;
+		contains = AR_CONTAINS(argv, 2);
+		return contains.longval == true;
+	case OP_STARTSWITH:
+		// String matching.
+		argv[0] = *aVal;
+		argv[1] = *bVal;
+		contains = AR_STARTSWITH(argv, 2);
+		return contains.longval == true;
```
Comment (jeffreylovitz) [inline src/filter_tree/filter_tree.c]: `contains.boolval`

File: src/filter_tree/filter_tree.c (hunk 8 above)
Comment (jeffreylovitz) [reply src/filter_tree/filter_tree.c]: Never mind, my bad!

**Pull request merged**